README.md
# haskell-aes

Implementation of the AES-128 encryption standard to practise my Haskell and to learn Stack/hspec.

Examples (which I used in testing) of correct AES-128 intermediate states are found in [this document](https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Standards-and-Guidelines/documents/examples/AES_Core128.pdf).

Currently supports ECB mode, but I will be extending it to CBC mode very soon.

**NOT SUITABLE FOR PRODUCTION USE.**


ChangeLog.md
# Changelog for haskell-aes

## Unreleased changes


test/Spec.hs
import SBoxSpec (runTests)
import PaddingSpec (runTests)
import RoundInternalSpec (runTests)
import RoundSpec (runTests)
import SubKeyGenSpec (runTests)
import EncryptDecryptSpec (runTests)

main :: IO ()
main = do
    SBoxSpec.runTests
    PaddingSpec.runTests
    RoundInternalSpec.runTests
    RoundSpec.runTests
    SubKeyGenSpec.runTests
    EncryptDecryptSpec.runTests


test/SBoxSpec.hs
module SBoxSpec
    ( runTests
    ) where

import Test.Hspec
import Test.QuickCheck
import Control.Exception (evaluate)

import SBox

runTests :: IO ()
runTests = hspec $
    describe "Sbox Module" $ do
        it "tests some specifc sBox outputs for correctness" $ do
            sBox 0x21 `shouldBe` 0xfd
            sBox 0xba `shouldBe` 0xf4
            sBox 0x5b `shouldBe` 0x39
            sBox 0x77 `shouldBe` 0xf5
        it "tests some specifc sBoxInv outputs for correctness" $ do
            sBoxInv 0xba `shouldBe` 0xc0
            sBoxInv 0x5c `shouldBe` 0xa7
            sBoxInv 0xe2 `shouldBe` 0x3b
            sBoxInv 0x0f `shouldBe` 0xfb
        it "ensures that sBox and sBoxInv are truly inverses" $
            property $ \x ->
                all ((==x) . ($x)) [sBoxInv . sBox, sBox . sBoxInv]


test/RoundSpec.hs
module RoundSpec
    ( runTests
    ) where

import Test.Hspec
import Test.QuickCheck
import Control.Exception (evaluate)

import Globals (Key, Block)
import Round

initialKeyAddOutput1 :: Block
initialKeyAddOutput1 = [0x40, 0xBF, 0xAB, 0xF4, 0x06, 0xEE, 0x4D, 0x30,
    0x42, 0xCA, 0x6B, 0x99, 0x7A, 0x5C, 0x58, 0x16]

subKey1 :: Key
subKey1 = [0xa0, 0xfa, 0xfe, 0x17, 0x88, 0x54, 0x2c, 0xb1,
    0x23, 0xa3, 0x39, 0x39, 0x2a, 0x6c, 0x76, 0x05]

roundOutput1 :: Block
roundOutput1 = [0xF2, 0x65, 0xE8, 0xD5, 0x1F, 0xD2, 0x39, 0x7B,
    0xC3, 0xB9, 0x97, 0x6D, 0x90, 0x76, 0x50, 0x5C]

runTests :: IO ()
runTests = hspec $ do
    describe "Round Module" $ do
        describe "roundEncrypt" $
            it "ensures that roundEncrypt produces correct output" $
                roundEncrypt subKey1 initialKeyAddOutput1 `shouldBe` roundOutput1
        describe "roundDecrypt" $
            it "ensures that roundDecrypt produces correct output" $
                roundDecrypt subKey1 roundOutput1 `shouldBe` initialKeyAddOutput1
        describe "check inverse holds" $
            it "checks that round encryption and decryption are inverses of each other" $ do
                encryptDecrypt initialKeyAddOutput1 `shouldBe` initialKeyAddOutput1
                decryptEncrypt initialKeyAddOutput1 `shouldBe` initialKeyAddOutput1
    where
        encryptDecrypt = roundDecrypt subKey1 . roundEncrypt subKey1
        decryptEncrypt = roundEncrypt subKey1 . roundDecrypt subKey1


app/Main.hs
module Main where

import System.Environment (getProgName, getArgs)
import System.Directory (doesFileExist)
import qualified Data.ByteString as B
import Data.Maybe (fromJust)
import Data.List.Split (chunksOf)
import Data.Char (toUpper)
import Numeric (readHex)

import Globals (Key, blockSize)
import EncryptDecrypt (encryptBlocksECB, decryptBlocksECB)

type Op = String
type Input = String
type Output = String
type KeyString = String

doOperation :: Op -> Input -> Output -> Key -> IO ()
doOperation "encrypt" input output key = do
    putStrLn "Reading input..."
    plaintextBlocks <- chunksOf blockSize . B.unpack <$> B.readFile input
    putStrLn "Encryping input..."
    ciphertextBlocks <- return . encryptBlocksECB key $ plaintextBlocks
    putStrLn "Writing to output file..."
    B.writeFile output . B.pack . concat $ ciphertextBlocks
doOperation "decrypt" input output key = do
    putStrLn "Reading input..."
    ciphertextBlocks <- chunksOf blockSize . B.unpack <$> B.readFile input
    putStrLn "Decryping input..."
    plaintextBlocks <- return . decryptBlocksECB key $ ciphertextBlocks
    case plaintextBlocks of
        Just blocks -> do
            putStrLn "Writing to output file..."
            B.writeFile output . B.pack . concat $ blocks
        Nothing ->
            putStrLn "Invalid data: unpad failed."

keyStringToKey :: KeyString -> Maybe Key
keyStringToKey ks 
    | length ks /= 32 = Nothing
    | allValidChars ks = Just key
    | otherwise = Nothing
        where
            key = map fst . concat . map (readHex) . chunksOf 2 $ ks
            allValidChars ks = all ((`elem` "0123456789ABCDEF") . toUpper) ks

parseKeyString :: KeyString -> IO (String, Maybe Key)
parseKeyString ks = case keyStringToKey ks of
    Nothing -> return ("Key is not a valid 128bit hex number\n", Nothing)
    k -> return ([], k)

checkArgs :: [String] -> IO ()
checkArgs [op, input, output, key] = do
    let opError = if elem op validOps then [] else opErrorMessage
    inputReal <- doesFileExist input
    let fileError = if inputReal then [] else input ++ " does not exist.\n"
    (keyError, parsedKey) <- parseKeyString key
    let errors = opError ++ fileError ++ keyError
    if length errors > 0 then
        putStr $ "Errors:\n\n" ++ errors
    else
        doOperation op input output $ fromJust parsedKey
            where
                validOps = ["encrypt", "decrypt"]
                opErrorMessage = "Operation must be either encrypt or decrypt\n"

main :: IO ()
main = do
    progName <- getProgName
    args <- getArgs
    if length args /= 4 then 
        putStrLn $ "Usage: " ++ progName ++ 
            "<encrypt/decrypt> <INPUT FILENAME> <OUTPUT FILENAME> <KEY>"
     else do
        checkArgs args


src/Padding.hs
module Padding
    ( pad,
      unpad,
    ) where

import Globals (Block, blockSize)

pad :: [Block] -> [Block]
pad [] = [generatePadding 0]
pad bs = init bs ++ (padded . last $ bs)
    where
        padded b
            | length b == blockSize = [b] ++ [(generatePadding 0)]
            | otherwise = [b ++ (generatePadding . length) b]

unpad :: [Block] -> Maybe [Block]
unpad [] = Nothing
unpad bs
    | last bs == generatePadding 0 = Just $ init bs
    | otherwise = (init bs ++) <$> (:[]) <$> (stripPadding . last) bs

generatePadding :: Int -> Block
generatePadding n = 0x80 : replicate (blockSize - 1 - n) 0x00

stripPadding :: Block -> Maybe Block
stripPadding b = let b' = dropWhile (==0) . reverse $ b in
    case b' of
        [] -> Nothing
        (0x80:ws) -> Just $ reverse ws
        otherwise -> Nothing


src/SubKeyGen.hs
module SubKeyGen
    ( generateSubKeys,
    ) where

import Control.Monad.State.Lazy
import Data.Word
import Data.Bits (xor)
import Data.List.Split (chunksOf)
import Crypto.Number.F2m (modF2m)

import Globals (Key, numRounds, aesPolynomial)
import SBox(sBox)

type KeyWord = [Word8]
type RC = Word8

generateSubKeys :: Key -> [Key]
generateSubKeys k = reverse keys
    where
        keys = snd . execState (keyGen numRounds k) $ (rcVals, [])

pushSubkey :: Key -> State ([RC], [Key]) ()
pushSubkey k = do
    (rcs, sks) <- get
    put (rcs, k:sks)

popRC :: State ([RC], [Key]) RC
popRC = do
    (rc:rcs, sks) <- get
    put (rcs, sks)
    return rc

keyGen :: Int -> Key -> State ([RC], [Key]) ()
keyGen 0 _ = return ()
keyGen n k = do
    rc <- popRC
    let w0' = w0 `mapXor` g rc w3
    let w1' = w0' `mapXor` w1
    let w2' = w1' `mapXor` w2
    let w3' = w2' `mapXor` w3
    let subKey = concat [w0', w1', w2', w3']
    pushSubkey subKey
    keyGen (n-1) subKey
        where
            mapXor = zipWith xor
            [w0, w1, w2, w3] = chunksOf 4 k

rcVals :: [RC]
rcVals = map (fromIntegral . modF2m aesPolynomial) (take numRounds $ doubles 1)
    where
        doubles n = n:(doubles $ 2 * n)

g :: RC -> KeyWord -> KeyWord
g rc kw = rcXor . map sBox $ kwRot
    where
        kwRot = let (l, r) = splitAt 1 kw in r ++ l
        rcXor (x:xs) = (rc `xor` x):xs


src/EncryptDecrypt.hs
module EncryptDecrypt
    ( encrypt,
      decrypt,
      encryptBlocksECB,
      decryptBlocksECB,
    ) where

import Control.Monad.State.Lazy
import Data.Bits (xor)

import Globals (Key, Block, numRounds)
import SubKeyGen (generateSubKeys)
import Round (roundEncrypt, roundDecrypt)
import Round.Internal (keyAdd, byteSub, byteSubInv, shiftRows, shiftRowsInv)
import Padding (pad, unpad)

encryptBlocksECB :: Key -> [Block] -> [Block]
encryptBlocksECB key = map (encrypt key) . pad

decryptBlocksECB :: Key -> [Block] -> Maybe [Block]
decryptBlocksECB key = unpad . map (decrypt key)

encrypt :: Key -> Block -> Block
encrypt key block = evalState state $ generateSubKeys key
    where
        state = encryptStateful numRounds $ keyAdd block key 

decrypt :: Key -> Block -> Block
decrypt key block = keyAdd key . evalState state $ reversedSubKeys
    where
        state = decryptStateful numRounds block
        reversedSubKeys = reverse $ generateSubKeys key

popSubKey :: State [Key] Key
popSubKey = do
    (subKey:ks) <- get
    put ks
    return subKey

encryptStateful :: Int -> Block -> State [Key] Block
encryptStateful 1 block = do
    subKey <- popSubKey
    return $ keyAdd subKey . shiftRows . byteSub $ block
encryptStateful n block = do
    subKey <- popSubKey
    encryptStateful (n-1) $ roundEncrypt subKey block

decryptStateful :: Int -> Block -> State [Key] Block
decryptStateful 0 block = return block
decryptStateful n block 
    | n == numRounds = do
        subKey <- popSubKey
        let finalRoundInv = byteSubInv . shiftRowsInv . keyAdd subKey
        decryptStateful (n-1) $ finalRoundInv block
    | n < numRounds = do
        subKey <- popSubKey
        decryptStateful (n-1) $ roundDecrypt subKey block


src/Round/Internal.hs
module Round.Internal
    ( keyAdd,
      byteSub,
      byteSubInv,
      shiftRows,
      shiftRowsInv,
      mixColumns,
      mixColumnsInv,
    ) where

import Data.Word (Word8)
import Data.Bits (xor)
import Data.List (transpose)
import Data.List.Split (chunksOf)
import Crypto.Number.F2m (addF2m, mulF2m)

import Globals (Key, Block, chunkSize, aesPolynomial)
import SBox (sBox, sBoxInv)

keyAdd :: Key -> Block -> Block
keyAdd = zipWith xor

byteSub :: Block -> Block
byteSub = map sBox

byteSubInv :: Block -> Block
byteSubInv = map sBoxInv

shiftRows :: Block -> Block
shiftRows = srCommon [0, 1, 2, 3]

shiftRowsInv :: Block -> Block
shiftRowsInv = srCommon [0, 3, 2, 1]

srCommon :: [Int] -> Block -> Block
srCommon shifts = concat . transpose . rotate . transpose . chunksOf chunkSize
    where
        rotate = zipWith rotateRow shifts
        rotateRow n xs = take (length xs) (drop n (cycle xs))

mixColumns :: Block -> Block
mixColumns = mcCommon matrixConsts
    where
        matrixConsts = [[2, 3, 1, 1], [1, 2, 3, 1], [1, 1, 2, 3], [3, 1, 1, 2]]

mixColumnsInv :: Block -> Block
mixColumnsInv = mcCommon matrixConsts
    where
        matrixConsts = [[0x0E, 0x0B, 0x0D, 0x09], [0x09, 0x0E, 0x0B, 0x0D],
            [0x0D, 0x09, 0x0E, 0x0B], [0x0B, 0x0D, 0x09, 0x0E]]

mcCommon :: [[Integer]] -> Block -> Block
mcCommon consts = map fromIntegral . concat . matrixMults . chunksOf chunkSize
    where
        matrixMults b = [zipWith doMults (repeat r) consts | r <- b]
        doMults l = foldr addF2m 0 . zipWith multWord l
        multWord = mulF2m aesPolynomial . fromIntegral


src/SBox.hs
module SBox
    ( sBox,
      sBoxInv,
    ) where

import Data.Word (Word8)

sBox :: Word8 -> Word8
sBox x = [0x63, 0x7C, 0x77, 0x7B, 0xF2, 0x6B, 0x6F, 0xC5,
    0x30, 0x01, 0x67, 0x2B, 0xFE, 0xD7, 0xAB, 0x76,
    0xCA, 0x82, 0xC9, 0x7D, 0xFA, 0x59, 0x47, 0xF0,
    0xAD, 0xD4, 0xA2, 0xAF, 0x9C, 0xA4, 0x72, 0xC0,
    0xB7, 0xFD, 0x93, 0x26, 0x36, 0x3F, 0xF7, 0xCC,
    0x34, 0xA5, 0xE5, 0xF1, 0x71, 0xD8, 0x31, 0x15,
    0x04, 0xC7, 0x23, 0xC3, 0x18, 0x96, 0x05, 0x9A,
    0x07, 0x12, 0x80, 0xE2, 0xEB, 0x27, 0xB2, 0x75,
    0x09, 0x83, 0x2C, 0x1A, 0x1B, 0x6E, 0x5A, 0xA0,
    0x52, 0x3B, 0xD6, 0xB3, 0x29, 0xE3, 0x2F, 0x84,
    0x53, 0xD1, 0x00, 0xED, 0x20, 0xFC, 0xB1, 0x5B,
    0x6A, 0xCB, 0xBE, 0x39, 0x4A, 0x4C, 0x58, 0xCF,
    0xD0, 0xEF, 0xAA, 0xFB, 0x43, 0x4D, 0x33, 0x85,
    0x45, 0xF9, 0x02, 0x7F, 0x50, 0x3C, 0x9F, 0xA8,
    0x51, 0xA3, 0x40, 0x8F, 0x92, 0x9D, 0x38, 0xF5,
    0xBC, 0xB6, 0xDA, 0x21, 0x10, 0xFF, 0xF3, 0xD2,
    0xCD, 0x0C, 0x13, 0xEC, 0x5F, 0x97, 0x44, 0x17,
    0xC4, 0xA7, 0x7E, 0x3D, 0x64, 0x5D, 0x19, 0x73,
    0x60, 0x81, 0x4F, 0xDC, 0x22, 0x2A, 0x90, 0x88,
    0x46, 0xEE, 0xB8, 0x14, 0xDE, 0x5E, 0x0B, 0xDB,
    0xE0, 0x32, 0x3A, 0x0A, 0x49, 0x06, 0x24, 0x5C,
    0xC2, 0xD3, 0xAC, 0x62, 0x91, 0x95, 0xE4, 0x79,
    0xE7, 0xC8, 0x37, 0x6D, 0x8D, 0xD5, 0x4E, 0xA9,
    0x6C, 0x56, 0xF4, 0xEA, 0x65, 0x7A, 0xAE, 0x08,
    0xBA, 0x78, 0x25, 0x2E, 0x1C, 0xA6, 0xB4, 0xC6,
    0xE8, 0xDD, 0x74, 0x1F, 0x4B, 0xBD, 0x8B, 0x8A,
    0x70, 0x3E, 0xB5, 0x66, 0x48, 0x03, 0xF6, 0x0E,
    0x61, 0x35, 0x57, 0xB9, 0x86, 0xC1, 0x1D, 0x9E,
    0xE1, 0xF8, 0x98, 0x11, 0x69, 0xD9, 0x8E, 0x94,
    0x9B, 0x1E, 0x87, 0xE9, 0xCE, 0x55, 0x28, 0xDF,
    0x8C, 0xA1, 0x89, 0x0D, 0xBF, 0xE6, 0x42, 0x68,
    0x41, 0x99, 0x2D, 0x0F, 0xB0, 0x54, 0xBB, 0x16] !! fromIntegral x

sBoxInv :: Word8 -> Word8
sBoxInv x = [0x52, 0x09, 0x6a, 0xd5, 0x30, 0x36, 0xa5, 0x38,
    0xbf, 0x40, 0xa3, 0x9e, 0x81, 0xf3, 0xd7, 0xfb,
    0x7c, 0xe3, 0x39, 0x82, 0x9b, 0x2f, 0xff, 0x87,
    0x34, 0x8e, 0x43, 0x44, 0xc4, 0xde, 0xe9, 0xcb,
    0x54, 0x7b, 0x94, 0x32, 0xa6, 0xc2, 0x23, 0x3d,
    0xee, 0x4c, 0x95, 0x0b, 0x42, 0xfa, 0xc3, 0x4e,
    0x08, 0x2e, 0xa1, 0x66, 0x28, 0xd9, 0x24, 0xb2,
    0x76, 0x5b, 0xa2, 0x49, 0x6d, 0x8b, 0xd1, 0x25,
    0x72, 0xf8, 0xf6, 0x64, 0x86, 0x68, 0x98, 0x16,
    0xd4, 0xa4, 0x5c, 0xcc, 0x5d, 0x65, 0xb6, 0x92,
    0x6c, 0x70, 0x48, 0x50, 0xfd, 0xed, 0xb9, 0xda,
    0x5e, 0x15, 0x46, 0x57, 0xa7, 0x8d, 0x9d, 0x84,
    0x90, 0xd8, 0xab, 0x00, 0x8c, 0xbc, 0xd3, 0x0a,
    0xf7, 0xe4, 0x58, 0x05, 0xb8, 0xb3, 0x45, 0x06,
    0xd0, 0x2c, 0x1e, 0x8f, 0xca, 0x3f, 0x0f, 0x02,
    0xc1, 0xaf, 0xbd, 0x03, 0x01, 0x13, 0x8a, 0x6b,
    0x3a, 0x91, 0x11, 0x41, 0x4f, 0x67, 0xdc, 0xea,
    0x97, 0xf2, 0xcf, 0xce, 0xf0, 0xb4, 0xe6, 0x73,
    0x96, 0xac, 0x74, 0x22, 0xe7, 0xad, 0x35, 0x85,
    0xe2, 0xf9, 0x37, 0xe8, 0x1c, 0x75, 0xdf, 0x6e,
    0x47, 0xf1, 0x1a, 0x71, 0x1d, 0x29, 0xc5, 0x89,
    0x6f, 0xb7, 0x62, 0x0e, 0xaa, 0x18, 0xbe, 0x1b,
    0xfc, 0x56, 0x3e, 0x4b, 0xc6, 0xd2, 0x79, 0x20,
    0x9a, 0xdb, 0xc0, 0xfe, 0x78, 0xcd, 0x5a, 0xf4,
    0x1f, 0xdd, 0xa8, 0x33, 0x88, 0x07, 0xc7, 0x31,
    0xb1, 0x12, 0x10, 0x59, 0x27, 0x80, 0xec, 0x5f,
    0x60, 0x51, 0x7f, 0xa9, 0x19, 0xb5, 0x4a, 0x0d,
    0x2d, 0xe5, 0x7a, 0x9f, 0x93, 0xc9, 0x9c, 0xef,
    0xa0, 0xe0, 0x3b, 0x4d, 0xae, 0x2a, 0xf5, 0xb0,
    0xc8, 0xeb, 0xbb, 0x3c, 0x83, 0x53, 0x99, 0x61,
    0x17, 0x2b, 0x04, 0x7e, 0xba, 0x77, 0xd6, 0x26,
    0xe1, 0x69, 0x14, 0x63, 0x55, 0x21, 0x0c, 0x7d] !! fromIntegral x
